kernel/boards/stm32f4_discovery/src/uart/mod.rs

```diff
@@ -0,0 +1,125 @@
+//! Implements UART specific functionality for the STM32F4 Discovery Board 
+use core::{convert::TryFrom, fmt};
+use irq_safety::MutexIrqSafe;
+use spin::Once;
+use stm32f4::stm32f407::{usart1, USART1, USART2, USART3, USART6, RCC, GPIOA};
+
+const USART1_BASE : *const usart1::RegisterBlock = USART1::ptr();
+const USART2_BASE : *const usart1::RegisterBlock = USART2::ptr();
+const USART3_BASE : *const usart1::RegisterBlock = USART3::ptr();
+const USART6_BASE : *const usart1::RegisterBlock = USART6::ptr();
+
+#[derive(Copy, Clone, Debug)]
+pub enum SerialPortAddress {
+    USART1,
+    USART2,
+    USART3,
+    USART6,
+}
+impl TryFrom<&str> for SerialPortAddress {
+	type Error = ();
+	fn try_from(s: &str) -> Result<Self, Self::Error> {
+		if s.eq_ignore_ascii_case("USART1") {
+			Ok(Self::USART1)
+		} else if s.eq_ignore_ascii_case("USART2") {
+			Ok(Self::USART2)
+		} else if s.eq_ignore_ascii_case("USART3") {
+			Ok(Self::USART3)
+		} else if s.eq_ignore_ascii_case("USART6") {
+			Ok(Self::USART6)
+		} else {
+			Err(())
+		}
+	}
+}
+
+static USART1_SERIAL_PORT: Once<MutexIrqSafe<SerialPort>> = Once::new();
+static USART2_SERIAL_PORT: Once<MutexIrqSafe<SerialPort>> = Once::new();
+static USART3_SERIAL_PORT: Once<MutexIrqSafe<SerialPort>> = Once::new();
+static USART6_SERIAL_PORT: Once<MutexIrqSafe<SerialPort>> = Once::new();
+
+
+/// Initialize UART for use.
+pub fn uart_init(uart_address: *const usart1::RegisterBlock) {
```

Comment:
a safe function cannot accept an arbitrary pointer and then dereference it. This either has to be marked as unsafe (which is undesirable IMO) or accept a different argument such as a reference to or an owned instance of the `RegisterBlock` type